Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean

```diff
@@ -0,0 +1,648 @@
+def lTensor.equiv_ofRightInverse {h : P → N} (hgh : Function.RightInverse h g) :
+    ((Q ⊗[R] N) ⧸ (LinearMap.range (lTensor Q f))) ≃ₗ[R] (Q ⊗[R] P) := {
+  lTensor.toFun Q hfg with
+  invFun    := lTensor.inverse_ofRightInverse Q hfg hgh
+  left_inv  := fun y ↦ by
+    simp only [lTensor.toFun, AddHom.toFun_eq_coe, coe_toAddHom]
+    obtain ⟨y, rfl⟩ := Submodule.mkQ_surjective _ y
+    simp only [Submodule.mkQ_apply, Submodule.liftQ_apply, lTensor.inverse_ofRightInverse_apply]
+  right_inv := fun z ↦ by
+    simp only [AddHom.toFun_eq_coe, coe_toAddHom]
+    obtain ⟨y, rfl⟩ := lTensor.surjective Q (hgh.surjective) z
+    rw [lTensor.inverse_ofRightInverse_apply]
+    simp only [lTensor.toFun, Submodule.liftQ_apply] }
+
+/-- For a surjective `f : N →ₗ[R] P`,
+  the natural equivalence between `Q ⊗ N ⧸ (image of ker f)` to `Q ⊗ P` -/
+noncomputable def lTensor.equiv :
+    ((Q ⊗[R] N) ⧸ (LinearMap.range (lTensor Q f))) ≃ₗ[R] (Q ⊗[R] P) :=
+  lTensor.equiv_ofRightInverse  Q hfg (Function.rightInverse_surjInv hg)
+
+/-- Tensoring an exact pair on the left gives an exact pair -/
+theorem lTensor_exact : Exact (lTensor Q f) (lTensor Q g) := by
+  rw [LinearMap.exact_iff]
+  rw [← Submodule.ker_mkQ (p := range (lTensor Q f))]
+  rw [← lTensor.inverse_comp_rTensor Q hfg hg]
+  apply symm
+  apply LinearMap.ker_comp_of_ker_eq_bot
+  rw [LinearMap.ker_eq_bot]
+  exact (lTensor.equiv Q hfg hg).symm.injective
+
+/-- Right-exactness of tensor product -/
+lemma lTensor_mkQ (N : Submodule R M) :
+    ker (lTensor Q (N.mkQ)) = range (lTensor Q N.subtype) := by
+  rw [← LinearMap.exact_iff]
+  exact lTensor_exact Q (LinearMap.exact_subtype_mkQ N) (Submodule.mkQ_surjective N)
+
+variable {M' N' P' : Type*}
+    [AddCommGroup M'] [AddCommGroup N'] [AddCommGroup P']
+    [Module R M'] [Module R N'] [Module R P']
+variable {f' : M' →ₗ[R] N'} {g' : N' →ₗ[R] P'}
+
+variable  (hfg' : Exact f' g') (hg' : Function.Surjective g')
+
+theorem TensorProduct.map_surjective : Function.Surjective (TensorProduct.map g g') := by
+  rw [← LinearMap.lTensor_comp_rTensor]
+  rw [LinearMap.coe_comp]
+  exact Function.Surjective.comp (lTensor.surjective _ hg') (rTensor.surjective _ hg)
+
+/-- Kernel of a product map (right-exactness of tensor product) -/
+theorem TensorProduct.map_ker :
+    LinearMap.ker (TensorProduct.map g g') =
+      LinearMap.range (lTensor N f') ⊔ LinearMap.range (rTensor N' f) := by
+  rw [← LinearMap.lTensor_comp_rTensor]
+  rw [LinearMap.ker_comp]
+  rw [← Exact.linearMap_ker_eq (rTensor_exact N' hfg hg)]
+  rw [← Submodule.comap_map_eq]
+  apply congr_arg₂ _ rfl
+  rw [LinearMap.range_eq_map, ← Submodule.map_comp, LinearMap.rTensor_comp_lTensor,
+    Submodule.map_top]
+  rw [← LinearMap.lTensor_comp_rTensor, LinearMap.range_eq_map, Submodule.map_comp,
+    Submodule.map_top]
+  rw [LinearMap.range_eq_top.mpr (rTensor.surjective M' hg), Submodule.map_top]
+  rw [Exact.linearMap_ker_eq (lTensor_exact P hfg' hg')]
+
+end Modules
+
+
+--  -·⬝.∙.⬝·-·⬝.∙.⬝·-
```

Comment:
Did you mean to leave this morse code here?

Reply:
Actually, this was not morse code, but me playing with the various dots that are all different to Lean…
I'll remove that line…